[WPE] Add build option ENABLE_WPE_LEGACY_API

carlosgcampos opened the pull request:
#### ba76fcf3a44613c08b18e3cb4ac72dd575682d16
<pre>
[WPE] Add build option ENABLE_WPE_LEGACY_API
<a href="https://bugs.webkit.org/show_bug.cgi?id=304237">https://bugs.webkit.org/show_bug.cgi?id=304237</a>

Reviewed by Alejandro G. Castro.

Enabled by default, but allows to build without the legacy api.

Canonical link: <a href="https://commits.webkit.org/304671@main">https://commits.webkit.org/304671@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a206d91080fa8d70ecb5896e9a8832b1483c7ae5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136253 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8608 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47532 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143963 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9291 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8453 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104181 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139198 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6755 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122092 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85012 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6411 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4070 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4556 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128210 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115700 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40293 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146708 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134737 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8292 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40860 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112523 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8308 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6966 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112866 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6338 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118397 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20994 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8340 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36454 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167516 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8058 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71899 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43706 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8280 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8132 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -158,6 +160,10 @@ include(GStreamerDependencies)
 WEBKIT_OPTION_END()
 
+if (NOT ENABLE_WPE_PLATFORM AND NOT ENABLE_WPE_LEGACY_API)
+    message(FATAL_ERROR "WPE Platform and Legacy API cannot be disabled at the same time")
```
Small nit: I think it is easier to parse if we worded this in a positive mode, and indicating which CMake option to toggle. For example, this could be:


```suggestion
    message(FATAL_ERROR "At least one of ENABLE_WPE_PLATFORM or ENABLE_WPE_LEGACY_API needs to be enabled")
```

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -924,6 +935,10 @@ static void webkitWebViewConstructed(GObject* object)
             priv->display = wpe_display_get_default();
         }
     }
+#else
```
This would leave out the WPEPlatform case when both `USE_LIBWPE` and `ENABLE_WPE_PLATFORM` are enabled. I think this should be:


```suggestion
#endif

#if ENABLE(WPE_PLATFORM)
```

carlosgcampos replied:
No, this is already inside a ENABLE(WPE_PLATFORM) block.

aperezdc commented on Source/WebKit/WebProcess/glib/WebProcessGLib.cpp:
```diff
@@ -216,6 +217,9 @@ void WebProcess::platformInitializeWebProcess(WebProcessCreationParameters& para
         } else
             initializePlatformDisplayIfNeeded();
     }
+#else
+    initializePlatformDisplayIfNeeded();
+#endif
```
Don't we want to keep the check for `!parameters.isServiceWorkerProcess` here? This is what I would have done:


```suggestion
    if (!parameters.isServiceWorkerProcess) {
#if USE(WPE_RENDERER)
        if (m_rendererBufferTransportMode.isEmpty()) {
            auto& implementationLibraryName = parameters.implementationLibraryName;
            if (!implementationLibraryName.isNull() && implementationLibraryName.data()[0] != '\0')
                wpe_loader_init(parameters.implementationLibraryName.data());
            PlatformDisplay::setSharedDisplay(PlatformDisplayLibWPE::create(parameters.hostClientFileDescriptor.release()));
        } else
#endif
            initializePlatformDisplayIfNeeded();
    }
```

carlosgcampos replied:
It would need more changes because isServiceWorkerProcess parameter is only defined for WPE_RENDERER. I wonder why this is only needed by WPE, I assumed this was added to not load any backend for service worker processes, that's why I left this specific to the old api.

aperezdc reviewed (commented):
I think there are a couple of locations where the guards are not completely correct. Please take a look at the comments&mdash;but please do comment if I am wrong about some of the guards, this is quite a long-winded patch and it's not always trivial to follow.

Also, a heads up for others reviewing this patch: we cannot replace `USE(LIBWPE) || ENABLE(WPE_PLATFORM)` with the simpler `PLATFORM(WPE)` because the PlayStation port enables `USE_LIBWPE` but of course uses `PLATFORM(PLATSTATION)` instead of `PLATFORM(WPE)`.

alexgcastro reviewed (approved):
LGTM! Great improvement being able to avoid libwpe.

Small suggestions, shall we rename the directory Tools/WebKitTestRunner/libwpe to wpe, and avoid compiling the files that are nto requires in the cmake. Maybe in a following patch we can do it, this is just big enough.

carlosgcampos committed:
[WPE] Add build option ENABLE_WPE_LEGACY_API
https://bugs.webkit.org/show_bug.cgi?id=304237

Reviewed by Alejandro G. Castro.

Enabled by default, but allows to build without the legacy api.

Canonical link: https://commits.webkit.org/304671@main
